d3-shape/index.d.ts

```diff
@@ -1,380 +1,2321 @@
+     */
+    padAngle(angle: (this: This, data: Datum[], ...args: any[]) => number): this;
 }
 
+/**
+ * Constructs a new pie generator with the default settings.
+ *
+ * Ensure that the accessors used with the pie generator correspond to the arguments passed into them,
+ * or set them to constants as appropriate.
+ */
 export function pie(): Pie<any, number | { valueOf(): number }>;
+/**
+ * Constructs a new pie generator with the default settings.
+ *
+ * Ensure that the accessors used with the pie generator correspond to the arguments passed into them,
+ * or set them to constants as appropriate.
+ *
+ * The generic refers to the data type of an element in the input array passed into the Pie generator.
+ */
 export function pie<Datum>(): Pie<any, Datum>;
+/**
+ * Constructs a new pie generator with the default settings.
+ *
+ * Ensure that the accessors used with the pie generator correspond to the arguments passed into them,
+ * or set them to constants as appropriate.
+ *
+ * The first generic corresponds to the type of the "this" context within which the pie generator and its accessor functions will be invoked.
+ *
+ * The second generic refers to the data type of an element in the input array passed into the Pie generator.
+ */
 export function pie<This, Datum>(): Pie<This, Datum>;
 
 // -----------------------------------------------------------------------------------
 // Line Generators
 // -----------------------------------------------------------------------------------
 
-
+/**
+ * The line generator produces a spline or polyline, as in a line chart.
+ * Lines also appear in many other visualization types, such as the links in hierarchical edge bundling.
+ *
+ * The generic refers to the data type of an element in the input array passed into the line generator.
+ */
 export interface Line<Datum> {
-    (data: Array<Datum>): string | undefined;
-    x(): (d: Datum, index: number, data: Array<Datum>) => number;
+    /**
+     * Generates a line for the given array of data. Depending on this line generator’s associated curve,
+     * the given input data may need to be sorted by x-value before being passed to the line generator.
+     *
+     * IMPORTANT: If the rendering context of the line generator is null,
+     * then the line is returned as a path data string.
+     *
+     * @param data Array of data elements.
+     */
+    (data: Datum[]): string | null;
+    /**
+     * Generates a line for the given array of data. Depending on this line generator’s associated curve,
+     * the given input data may need to be sorted by x-value before being passed to the line generator.
+     *
+     * IMPORTANT: If the line generator has been configured with a rendering context,
+     * then the line is rendered to this context as a sequence of path method calls and this function returns void.
+     *
+     * @param data Array of data elements.
+     */
+    (data: Datum[]): void;
```

Comment:
I don't think this overload will work -- there's nothing in the parameter signature that will allow TypeScript to choose it. Unfortunately, I think you'll have to add an `| undefined` at the end of the first signature.